src/NuGet.Services.V3/CommitCollectorUtility.cs

```diff
@@ -0,0 +1,170 @@
+
+                if (entry.CommitTimeStamp > max.CommitTimeStamp)
+                {
+                    max = entry;
+                }
+                else if (entry.CommitTimeStamp == max.CommitTimeStamp)
+                {
+                    const string message = "There are multiple catalog leaves for a single package at one time.";
+                    _logger.LogError(
+                        message + " ID: {PackageId}, version: {PackageVersion}, commit timestamp: {CommitTimestamp:O}",
+                        entry.PackageIdentity.Id,
+                        entry.PackageIdentity.Version.ToFullString(),
+                        entry.CommitTimeStamp);
+                    throw new InvalidOperationException(message);
+                }
+            }
+
+            return max;
+        }
+
+        public ConcurrentBag<IdAndValue<IReadOnlyList<CatalogCommitItem>>> GroupById(List<CatalogCommitItem> latestItems)
+        {
+            var workEnumerable = latestItems
+                .GroupBy(x => x.PackageIdentity.Id, StringComparer.OrdinalIgnoreCase)
+                .Select(x => new IdAndValue<IReadOnlyList<CatalogCommitItem>>(x.Key, x.ToList()));
+
+            var allWork = new ConcurrentBag<IdAndValue<IReadOnlyList<CatalogCommitItem>>>(workEnumerable);
+            return allWork;
+        }
+
+        public async Task<IReadOnlyDictionary<CatalogCommitItem, PackageDetailsCatalogLeaf>> GetEntryToLeafAsync(
+            IEnumerable<CatalogCommitItem> entries)
+        {
+            var packageDetailsEntries = entries.Where(IsOnlyPackageDetails);
+            var allWork = new ConcurrentBag<CatalogCommitItem>(packageDetailsEntries);
+            var output = new ConcurrentBag<KeyValuePair<CatalogCommitItem, PackageDetailsCatalogLeaf>>();
+
+            using (_telemetryService.TrackCatalogLeafDownloadBatch(allWork.Count))
+            {
+                var tasks = Enumerable
+                    .Range(0, _options.Value.MaxConcurrentCatalogLeafDownloads)
+                    .Select(async x =>
+                    {
+                        await Task.Yield();
+                        while (allWork.TryTake(out var work))
+                        {
+                            try
+                            {
+                                _logger.LogInformation(
+                                    "Downloading catalog leaf for {PackageId} {Version}: {Url}",
+                                    work.PackageIdentity.Id,
+                                    work.PackageIdentity.Version.ToNormalizedString(),
+                                    work.Uri.AbsoluteUri);
+
+                                var leaf = await _catalogClient.GetPackageDetailsLeafAsync(work.Uri.AbsoluteUri);
+                                output.Add(KeyValuePair.Create(work, leaf));
+                            }
+                            catch (Exception ex)
+                            {
+                                _logger.LogError(
```

Comment:
The latest dev branch has 2.2 logger. If you pull that in you should be able to drop the `0`.

Reply:
Done.